design/TextSearch.md

```diff
@@ -0,0 +1,49 @@
+# Text Search Design
+
+(aka "BigSearch")
+
+## Key ideas
+
+- Users create text files that provide text data to index and search.
+- Search indexing builds internal resources for faster query results.
+- Each search query leads to zero or more search matches.
+- Each search match includes an internal file position.
+- Files permit queries for data surrounding each match (each internal file position).
+
+### File
+
+- Uniquely-identified resource.
+- All content is text (for now).
+- CRUD API for creating, reading and updating files.
+
+#### File data model
+
+ - Each file is a Sequence of text.
+ - Each file is potentially very large in size.
+ - Use BigMap to store file chunks, if needed.
+
+### Search
+
+Concepts:
+
+- Indexing
+- Matches
+- Queries
+
+#### Search indexing
+
+ - Indexing constructs structures that accelerate search
```

Comment:
What structures are you thinking about here? I remember implementing a [Suffix Tree](https://en.wikipedia.org/wiki/Suffix_tree) for some programming exercise a long time ago, and reading that they'd be suited to this kind of problem.

Reply:
In the limit, I want to represent very large sequences (much longer than a single word or phrase) and permit fast edits, no matter how randomly distributed those edits may be (imagine they are coming from a firehose of concurrent editors all banging on the same shared sequence).

In the short term, the features that the `motoko-sequence` library has today suffice.  More here: https://github.com/matthewhammer/motoko-bigtext/pull/2#discussion_r485160713